more_itertools/more.py

```diff
@@ -1591,18 +1591,43 @@ def sort_together(iterables, key_list=(0,), reverse=False):
         >>> sort_together(iterables, key_list=(1, 2))
         [(2, 3, 1), (0, 0, 1), ('a', 'c', 'b')]
 
+    To sort by a function of the elements of the iterable, pass a *key*
+    function. Its arguments are the elements of the iterables corresponding to
+    the key list::
+
+        >>> iterables = [(3, 1, 2), (1, 0, 2), ('c', 'b', 'a')]
+        >>> sort_together(iterables, key_list=(0, 1), key=lambda x, y: x - y)
+        [(2, 1, 3), (2, 0, 1), ('a', 'b', 'c')]
+
     Set *reverse* to ``True`` to sort in descending order.
 
         >>> sort_together([(1, 2, 3), ('c', 'b', 'a')], reverse=True)
         [(3, 2, 1), ('a', 'b', 'c')]
 
     """
-    return list(
-        zip(
-            *sorted(
-                zip(*iterables), key=itemgetter(*key_list), reverse=reverse
+    if key is None:
+        # if there is no key function, the key argument to sorted is an
+        # itemgetter
+        key_argument = itemgetter(*key_list)
+    else:
+        # if there is a key function, call it with the items at the offsets
+        # specified by the key function as arguments
+        key_list = list(key_list)
+        if len(key_list) == 1:
```

Comment:
Thanks for breaking this out. I'm still a little puzzled - does the length-1 case actually fail and need special handling? If so, why?

Reply:
If we were to use the code from the other case, the `itemgetter()` would return a single element (rather than a tuple), and when we try to pass them to the key function with `*` (iterable unpacking) it would raise a `TypeError` that "argument after * must be an iterable".

The essence of the problem is that the degenerate case of `itemgetter()` with a single argument returns a single value, whereas when given multiple arguments it returns a tuple.

Reply:
OK, got it - thanks.

```python
>>> from operator import itemgetter
... 
... get_1 = itemgetter(1)
... get_1([1, 2, 3])
... 
... get_1_2 = itemgetter(1, 2)
... get_1_2([1, 2, 3])
1: (2, 3)
```